Path support

erooke opened the pull request:
For all import / export methods make the path a `Union[Pathlike, str, bytes]` argument to follow python conventions as outlined in [pep 519](https://peps.python.org/pep-0519/).

closes: https://github.com/gumyr/build123d/issues/685
closes: https://github.com/gumyr/build123d/issues/774

~~I got carried away big time and this PR is probably unreasonably large. To test the new functionality I changed all the tests to feed in `pathlib.Paths`. In addition any tests which wrote to disk were rewritten to use the pytest `tmp_path` fixture. Which then caused me to rewrite the tests in the pytest style...~~

~~I can revert the test changes and be less invasive with my changes if desired.~~

Edit: I decided to roll those back. I think they're good changes in the long run but they get in the way of reviewing the pr at hand.

Open questions:
- should the deprecated export methods in `topology.py` also be moved over to this scheme? Technically that is what #685 is requesting
- ~~should the tests attempt to verify that all three types: string, paths, and bytes work?~~ I added said tests 
- did I miss any exporting/importing methods

erooke committed:
test: 3d exporters work with all types of paths

erooke committed:
test: 2d exporters work with pathlike options

erooke committed:
test: importers work with all pathlike arguments

erooke committed:
test: mesher works with pathlike objects

erooke committed:
feat: allow mesher to read/write from pathlike things

gumyr commented on src/build123d/exporters.py:
```diff
@@ -1428,13 +1428,13 @@ def _color_attribs(c: Color) -> Tuple[str, str]:
 
         Args:
-            path (str): The file path where the SVG data will be written.
+            path (Union[str, Path]): The file path where the SVG data will be written.
```
`bytes` is missing here

erooke replied:
good catch

gumyr commented on src/build123d/exporters3d.py:
```diff
@@ -153,25 +153,27 @@ def _create_xde(to_export: Shape, unit: Unit = Unit.MM) -> TDocStd_Document:
 
 
-def export_brep(to_export: Shape, file_path: Union[str, BytesIO]) -> bool:
+def export_brep(
```
`BytesIO` provides a mechanism for in memory data exchange which seems to be dropped here, or at least not visible.

erooke replied:
Pushed a fix and added tests to make sure the next guy doesn't accidentally clobber that behavior too :sweat_smile:

gumyr commented on src/build123d/exporters3d.py:
```diff
@@ -349,9 +351,8 @@ def export_stl(
-        writer.ASCIIMode = True
-    else:
-        writer.ASCIIMode = False
+    writer.ASCIIMode = ascii_format
```
Yikes .. thanks for fixing this!

gumyr reviewed (changes requested):
For my education, what does the input type `bytes` actually do here? Why would a user pass `bytes` - is it for unicode file names?

gumyr commented:
There is no need to refactor the deprecated export methods.

erooke committed:
feat: support Path objects for import / export

For all import / export methods make the path a `Union[Pathlike, str, bytes]`
argument to follow python conventions as outlined in pep 519

erooke committed:
test: export_brep works with BytesIO

erooke committed:
fix: export_brep works with BytesIO

erooke commented:
> For my education, what does the input type `bytes` actually do here? Why would a user pass `bytes` - is it for unicode file names?

I am not 100% sure. It seems to be a leftover from the fact that the c standard library used raw byte strings to represent paths. There is some discussion in pep 519 [here](https://peps.python.org/pep-0519/#have-fspath-only-return-strings). I included it in the signature because all of the underlying `Pathlike` functions support it and they claim:

> In the end, it was decided that using bytes to represent paths is simply not going to go away and thus they should be supported to some degree. The hope is that people will gravitate towards path objects like pathlib and that will move people away from operating directly with bytes.

gumyr commented:
Thank you!  Another nice improvement.

DeflateAwning commented:
Thank you! I've been looking forward to this very small quality-of-life improvement :)
